trunk/source/Library-Proofs/src/de/uni_freiburg/informatik/ultimate/lib/proofs/floydhoare/FloydHoareUtils.java

```diff
@@ -127,6 +127,29 @@ public static void createInvariantResults(final String pluginName, final IIcfg<I
 			}
 			new WitnessInvariant(invResult.getInvariant()).annotate(locNode);
 		}
+		
+		// Temporary hack: Get the invariant of the pre-error location
+		// There exists only one in our use case, needs to be generalized if obtaining
+		// invariants for all pre-error locations is desired
+		final var errorLocs = IcfgUtils.getErrorLocations(icfg).iterator();
+		if (errorLocs.hasNext())
```

Comment:
The easiest and cleanest solution to me seems to be to write a small method (here, or perhaps even in `IcfgUtils`) which collects all predecessors of error locations in a `Set<IcfgLocation>`, and then just add them to `locsForLoopLocations` (just like `getPotentialCycleProgramPoints` above).

At that point, we could then rename the set to sth like `locsForInvariants` ;-)

Reply:
Agree, this seems like the best solution. In our setting, we only have a single unique predecessor for all error locations whose invariant we utilize. That means for each requirement check, we do something with the resulting invariant of said location. There's not really any combining involved as it is obscuring the results. However, I don't think that changing the code to handle all predecessors of error locations would change anything for our setting, so a cleaner implementation like that can easily be done while still being in agreement to our use case, at least if I understand correctly.

Reply:
I have implemented a function in IcfgUtils that returns the predecessors of all error locations in b4ff0a6

After looking more into the code, the issue is more complicated than I initially thought, however. I try to explain in a short manner:

- In our redundancy analysis, assuming we look at N requirements, we have N error locations. While conceptually it is true that there is only one predecessor, in truth there is a predecessor for each of the N error locations. However, all of the predecessors will always have the same location invariant in our setting. Hence, we only need a single invariant result from any of these predecessor locations.

- The way it is coded now, we create an invariant result for all predecessors. What happens in Pea2Boogie is the following: we have a method that takes trace abstraction results and transforms them into our desired results. (Now, if we check the redundancy of let's say a requirement called r1 with a total of N = 1000 requirements, and it turns out that r1 is indeed redundant, trace abstraction will throw 1000 location invariant results for the predecessors of the error locations into this method. However, all of them will have the same location invariant, which we only want to receive once to output it. https://github.com/ultimate-pa/ultimate/blob/c80803b6d3280088281ea55be03e987d0dfadddb/trunk/source/PEAtoBoogie/src/de/uni_freiburg/informatik/ultimate/pea2boogie/VerificationResultTransformer.java#L130 https://github.com/ultimate-pa/ultimate/blob/c80803b6d3280088281ea55be03e987d0dfadddb/trunk/source/PEAtoBoogie/src/de/uni_freiburg/informatik/ultimate/pea2boogie/PEAtoBoogieObserver.java#L99

- To circumvent the issue, I added a temporary set into the VerificationResultTransformer until I can think of a better way to handle this. This way the same location invariant is not output N times but only once per requirement that is redundant. Though, the issue remains as all of the 1000 invariant results still enter the method, and 999 are then rejected which is just a waste and not acceptable if the possibility exists to just create one for our use case. In order to think of a better solution I have the following questions:

1. Would it be possible to have settings for the invariants one wants to create? 
2. Is it possible to check the settings of a project in a different project?
3. Do you maybe have an idea how we could fit the need of the use case described?

Reply:
1. I'm planning to add something like this. We should discuss your use case in more detail (preferably in person) so we can figure out how to best design such settings.
2. We should not read settings of other projects.
3. If I understand correctly, you start one CEGAR loop per error location, and you always want the invariant at the predecessor of that particular error location that the CEGAR loop was started for? I think this is possible as follows:
   - In `FloydHoareUtils::createInvariantResults` you have access to the checked specification as `spec = annotation.getSpecification()`
   - From this specification, you should be able to identify which error location's unreachability the CEGAR loop has proven. Just filter the error locations by `spec::isFinalState`. Have a look at `FloydHoareUtils::getCheckedSpecifications` for an example.